hasktorch/src/Torch/Typed/NN/Recurrent/LSTM.hs

```diff
@@ -0,0 +1,478 @@
+        -> ParamsPerDirection device dtype inputDim hiddenSize numLayers 'Unidirectional
+
+
+sampleBidirectionalParams
+    :: forall device dtype inputDim hiddenDim numLayers
+     . ( KnownNat hiddenDim
+       , KnownNat inputDim
+       , KnownNat numLayers
+       , KnownDType dtype
+       , LayerSong numLayers
+       , KnownDevice device
+       , RandDTypeIsValid device dtype
+       )
+    => IO
+           ( ParamsPerDirection
+                 device
+                 dtype
+                 inputDim
+                 hiddenDim
+                 numLayers
+                 'Bidirectional
+           )
+sampleBidirectionalParams =
+    BidirectionalParams <$> singLayerSing <*> singLayerSing
+
+
+sampleUniDirectionalParams
+    :: ( KnownNat hiddenDim
+       , KnownNat inputDim
+       , KnownNat numLayers
+       , KnownDType dtype
+       , LayerSong numLayers
+       , KnownDevice device
+       , RandDTypeIsValid device dtype
+       )
+    => IO
+           ( ParamsPerDirection
+                 device
+                 dtype
+                 inputDim
+                 hiddenDim
+                 numLayers
+                 'Unidirectional
+           )
+sampleUniDirectionalParams = UniDirectionalParams <$> singLayerSing
+
+instance A.Parameterized  (ParamsPerDirection device dtype inputDim hiddenDim numLayers dir) where
+    flattenParameters (UniDirectionalParams ps) = A.flattenParameters ps
+    flattenParameters (BidirectionalParams as bs) =
+        A.flattenParameters as <> A.flattenParameters bs
+    replaceOwnParameters (UniDirectionalParams ps) =
+        UniDirectionalParams <$> A.replaceOwnParameters ps
+    replaceOwnParameters (BidirectionalParams as bs) =
+        BidirectionalParams
+            <$> A.replaceOwnParameters as
+            <*> A.replaceOwnParameters bs
+
+-- | A long, short-term memory layer with either fixed initial
+-- states for the memory cells and hidden state or learnable
+-- inital states for the memory cells and hidden state.
+--
+data LSTM device (dir :: Directionality)
+            (dtype :: D.DType)
+            (numLayers :: Nat)
+            (inputDim :: Nat)
+            (hiddenDim :: Nat) =  LSTM {
+      lstm_params :: ParamsPerDirection device dtype inputDim hiddenDim numLayers dir
+      , lstm_dropout :: Dropout
+      , lstm_init_c     :: Tensor device dtype '[numLayers * (NumberOfDirections dir), hiddenDim]
+      , lstm_init_h     :: Tensor device dtype '[numLayers * (NumberOfDirections dir), hiddenDim]
+    }
+    | LSTMLearnedInit {
+        lstmLearnedInit_params :: ParamsPerDirection device dtype inputDim hiddenDim numLayers dir
+        , lstmLearnedInit_dropout :: Dropout
+        , lstmLearnedInit_init_c     :: Parameter device dtype '[numLayers * (NumberOfDirections dir), hiddenDim]
+        , lstmLearnedInit_init_h     :: Parameter device dtype '[numLayers * (NumberOfDirections dir), hiddenDim]
+    }
+    deriving Generic
```

Comment:
that's cool and gives a lot of flexibility. thanks for including these two options here